src/containers/VerletListHelpers.h

```diff
@@ -45,7 +45,102 @@ class VerletListHelpers {
         // (ensured by traversals)
         // also the list is not allowed to be resized!
 
-        _verletListsAoS[&i].push_back(&j);
+        _verletListsAoS.at(&i).push_back(&j);
+    }
+
+    virtual void SoAFunctor(SoA &soa, bool /*newton3*/ = true) override {
+      if (soa.getNumParticles() == 0) return;
+
+      Particle **const __restrict__ idptr = reinterpret_cast<Particle **const>(soa.begin(Particle::AttributeNames::id));
+      double *const __restrict__ xptr = soa.begin(Particle::AttributeNames::posX);
+      double *const __restrict__ yptr = soa.begin(Particle::AttributeNames::posY);
+      double *const __restrict__ zptr = soa.begin(Particle::AttributeNames::posZ);
+
+      size_t numPart = soa.getNumParticles();
+      for (unsigned int i = 0; i < numPart; ++i) {
+        auto &currentList = _verletListsAoS.at(idptr[i]);
+
+        for (unsigned int j = i + 1; j < numPart; ++j) {
+          const double drx = xptr[i] - xptr[j];
+          const double dry = yptr[i] - yptr[j];
+          const double drz = zptr[i] - zptr[j];
+
+          const double drx2 = drx * drx;
+          const double dry2 = dry * dry;
+          const double drz2 = drz * drz;
+
+          const double dr2 = drx2 + dry2 + drz2;
+
+          if (dr2 < _cutoffskinsquared) {
+            currentList.push_back(idptr[j]);
+          }
+        }
+      }
+    }
+
+    virtual void SoAFunctor(SoA &soa1, SoA &soa2, bool /*newton3*/ = true) override {
+      if (soa1.getNumParticles() == 0 || soa2.getNumParticles() == 0) return;
+
+      Particle **const __restrict__ id1ptr =
+          reinterpret_cast<Particle **const>(soa1.begin(Particle::AttributeNames::id));
+      double *const __restrict__ x1ptr = soa1.begin(Particle::AttributeNames::posX);
+      double *const __restrict__ y1ptr = soa1.begin(Particle::AttributeNames::posY);
+      double *const __restrict__ z1ptr = soa1.begin(Particle::AttributeNames::posZ);
+
+      Particle **const __restrict__ id2ptr =
+          reinterpret_cast<Particle **const>(soa2.begin(Particle::AttributeNames::id));
+      double *const __restrict__ x2ptr = soa2.begin(Particle::AttributeNames::posX);
+      double *const __restrict__ y2ptr = soa2.begin(Particle::AttributeNames::posY);
+      double *const __restrict__ z2ptr = soa2.begin(Particle::AttributeNames::posZ);
+
+      size_t numPart1 = soa1.getNumParticles();
+      for (unsigned int i = 0; i < numPart1; ++i) {
+        auto &currentList = _verletListsAoS.at(id1ptr[i]);
+
+        size_t numPart2 = soa2.getNumParticles();
+
+        for (unsigned int j = 0; j < numPart2; ++j) {
```

Comment:
same as above

Reply:
there is a function call inside here: `currentList.push_back(...)`, so a vectorization is not possible immediately.
I have actually tried to do a vectorization here by allocating an additional array and just marking the respective fields. This was vectorizable, but slower than the method in here.